### ISSUE TITLE
`OnPollStatusChange` for `pallet-referenda`

### ISSUE THREAD

**Issue by pandres95** [paritytech/polkadot-sdk#4227]
### Is there an existing issue?

- [X] I have searched the existing issues

### Experiencing problems? Have you tried our Stack Exchange first?

- [X] This is not a support question.

### Motivation

For pallets that use `pallet_referenda` as `Polling` implementor might be benefitted with having knowledge of the change of a poll status (i.e. for releasing locks of `Completed` polls)

### Request

Implement a trait that allows pallets to listen for the poll status change event and act accordingly.

### Solution

Such trait might have the following structure, where `poll_status` is a reference to alllow implement for tuples without `Clone` constraints:

```rs
pub trait OnPollStatusChange<Tally, Moment, Class, Index: Copy> {
	fn on_poll_status_change(index: Index, poll_status: &PollStatus<Tally, Moment, Class>);
}
```

### Are you willing to help with this request?

Yes!

**Comment by kianenigma** [comment]
I am a bit worried that wanting to clear up a lot of votes in a synchronous way is more of a footgun. What if a lot of votes are suddenly eligible for being unlocked in one go? 

I think such things are best done async. This one seems like a good use case of pallet tasks or on_idle. @gupnik wdyt? 

I am not against having some hooks in the referenda pallet per se.

**Comment by gupnik** [comment]
> I think such things are best done async. This one seems like a good use case of pallet tasks or on_idle. @gupnik wdyt?

Indeed. As @olanod mentioned [here](https://github.com/paritytech/polkadot-sdk/pull/4228#issuecomment-2082284602), it would be better to keep it extremely light weight and then use the Tasks API to perform the cleanup. Here's an example that might help: https://github.com/paritytech/polkadot-sdk/blob/f34d8e3cf033e2a22a41b505c437972a5dc83d78/substrate/frame/examples/tasks/src/lib.rs#L59

**Comment by pandres95** [comment]
I have some thoughts about some of the comments @kianenigma raised:

> I think such things are best done async. This one seems like a good use case of pallet tasks or on_idle.

Completely agree! The aim behind this issue is precisely to keep things as lightweight as possible. How we came up with the idea of proposing the hooks is the case where we didn't have it and we had to use `on_idle` to try to iterate over every closed poll we still had voting records for, then clear the votes using `CommunityVotes::clear_prefix(number_of_votes)`.

Instead, having the notification/hook, we don't have to go through all these expensive iterations.

> What if a lot of votes are suddenly eligible for being unlocked in one go?

What we contemplate in our use case, which is a bit different from the one implemented at `pallet-conviction-voting` is that votes are decoupled from locks. So, in the case we clear up votes from storage, that doesn't mean locks are being removed. That is a process that should happen manually, called by the voter via [an extrinsic](https://github.com/virto-network/virto-node/blob/master/pallets/communities/src/lib.rs#L528).

> I am not against having some hooks in the referenda pallet per se.

I thought about your proposition of granular notifications and everyone's concerns about spending block weight, and have though about changing the design of the hook to deliver notifications at the end of the block, using `on_idle`. This would imply temporarily storing the poll changes that happen throughout the block, and clearing them up once they've been notified.

```rs
	#[pallet::hooks]
	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {

		fn on_idle(now: BlockNumberFor<T>, limit: Weight) -> Weight {
			let mut meter = WeightMeter::with_limit(limit);

			if meter.try_consume(Self::on_idle_weight()).is_err() {
				log::debug!(target: LOG, "Not enough weight for on_idle. {} < {}", Self::on_idle_weight(), limit);
				return meter.consumed()
			}

			// ...
			for (poll_index, poll_change) of PollChangesToNotify::<T, I>::iter() {
				match poll_change {
					PollChange::None => T::OnPollStatusChange::notify_poll_none(poll_index),
					PollChange::Ongoing(tally, track) => T::OnPollStatusChange::notify_poll_ongoing(poll_index, tally, track),
					PollChange::Completed(when, result) => T::OnPollStatusChange::notify_poll_completed(poll_index, when, result)
				}
			}

			meter.consumed()
		}
	}
```

Now, by having the granular hooks dispatched when `on_idle`, and the Tasks API (kudos to @gupnik for the idea), we could come up with something way too simple as this:

```rs
impl<T> OnPollStatusChange<pallet_communities::TallyOf<T>, BlockNumberFor<T>, CommunityIdOf<T>, u32>
	for Pallet<T>{
	fn notify_poll_completed(poll_index: u32, _when: BlockNumberFor<T>, _result: bool) {
		frame_system::do_task(Task::<T>::ClearVotes { poll_index }.into());
	}
}
```

What are your thoughts?

**Comment by gupnik** [comment]
> Now, by having the granular hooks dispatched when `on_idle`, and the Tasks API (kudos to @gupnik for the idea), we could come up with something way too simple as this:
> 
> ```rust
> impl<T> OnPollStatusChange<pallet_communities::TallyOf<T>, BlockNumberFor<T>, CommunityIdOf<T>, u32>
> 	for Pallet<T>{
> 	fn notify_poll_completed(poll_index: u32, _when: BlockNumberFor<T>, _result: bool) {
> 		frame_system::do_task(Task::<T>::ClearVotes { poll_index }.into());
> 	}
> }
> ```
> 
> What are your thoughts?

I do want to clarify that this won't be the correct way to use the tasks here, as you still end up using the task weight for processing during this call. Instead, you would want to trigger them using something like an off-chain worker. The notification here should only modify the condition to obtain valid tasks.

**Comment by pandres95** [comment]
Ah, I see! So, tasks should be passively invoked, either via a extrinsic, or via a off-chain worker? A third alternative would be scheduling an extrinsic on a next block (the way that, for example, `pallet-referenda` does).

Either case, I still think there are valid use cases for getting the poll status change notifications, as long as notification observers don't alter weight extensively.

Finally, I think another modification that could be done in the notifiers is that they return the used weight the same way `on_idle` would do. Something similar to this:

```rs
impl<T> OnPollStatusChange<pallet_communities::TallyOf<T>, BlockNumberFor<T>, CommunityIdOf<T>, u32>
	for Pallet<T>{
	fn notify_poll_completed(
		poll_index: u32,
		_as_completed: (BlockNumberFor<T>, bool),
		limit: Weight
	) -> Weight {
		let mut meter = WeightMeter::with_limit(limit);

		if meter.try_consume(Self::on_idle_weight()).is_err() {
			log::debug!(target: LOG, "Not enough weight for on_idle. {} < {}", Self::on_idle_weight(), limit);
			return meter.consumed()
		}

		frame_system::do_task(Task::<T>::ClearVotes { poll_index }.into());
		
		meter.consumed()
	}
}
```

I'm not really confident about this approach. However, if this strategy can be used to avoid exhausting weight limits, it's fine.

**Comment by Polkadot-Forum** [comment]
This issue has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/2024-05-21-technical-fellowship-opendev-call/8264/1
